### PR TITLE
Add `Enumerable#first` with fallback block

### PR DESCRIPTION
**Pull request by MakeNowJust**
Such a method exists in `Indexable`, but not in `Enumerable`.
It simplify `first`/`first?` implementation a bit.

### CODE REVIEW

**Comment by MakeNowJust** [comment]
`Indexable#first` and `#first?` are removed. Then, `#first` raises `Enumerable#EmptyError` now. Is this breaking-change?

**Comment by RX14** [comment]
Unifying enumerable and indexable exceptions sounds something worth considering

**Comment by MakeNowJust** [comment]
Ah, I'll fix `Deque` spec to pass CI.

File: src/indexable.cr [hunk 1]
```diff
@@ -454,16 +444,6 @@ module Indexable(T)
     size == 0 ? yield : unsafe_fetch(0)
```
Comment (bcardiff) [inline src/indexable.cr:444]: Might be worth moving it the examples to enumerable. Even if the use arrays (and show one that raises `Enumerable::EmptyError`)

Then the `:inherited:` can be used. Or, maybe this `def first` can also go. Is not that the default implementation will be more inneficient.

**Comment by MakeNowJust** [comment]
Docs are updated and specs are fixed.

File: src/enumerable.cr [hunk 2]
```diff
@@ -479,18 +479,38 @@ module Enumerable(T)
     if_none
   end
 
+  # Returns the first element in the collection,
+  # or the given block's value.
```
Comment (straight-shoota) [inline src/enumerable.cr:483]:
```suggestion
  # If the collection is empty, calls the block and returns its value.
```

**Comment by straight-shoota** [comment]
@bcardiff Alternatively move `IndexError` to `Indexable::IndexError`? Both errors are strongly related to the respective collection types.

**Comment by MakeNowJust** [comment]
I think `Array#pop` and `#shift` can raise `EmptyError` instead `IndexError`.

But I'm not sure this discussion is scope of this PR.

**Pull request merged**